include/swift/Runtime/Mutex.h

```diff
@@ -186,121 +114,687 @@ class Mutex {
   /// - The behavior is undefined if the calling thread already owns
   ///   the mutex (likely a deadlock)?
   /// - Does not throw exceptions but will halt on error (fatalError).
-  bool try_lock() { return Impl.try_lock(); }
+  bool try_lock() { return MutexPlatformHelper::try_lock(Handle); }
 
-public:
   /// Releases lock, waits on supplied condition, and relocks before returning.
   ///
-  /// Precondition: Mutex locked by this thread, undefined otherwise.
-  void wait(Condition &condition) { Impl.wait(condition); }
+  /// Precondition: Mutex held by this thread, undefined otherwise.
+  void wait(ConditionVariable &condition) {
+    ConditionPlatformHelper::wait(condition.Handle, Handle);
+  }
 
-public:
-  /// Acquires lock before calling the supplied critical section and release
+  /// Acquires lock before calling the supplied critical section and releases
   /// lock on return from critical section.
   ///
+  /// This call can block while waiting for the lock to become available.
+  ///
   /// For example the following mutates value while holding the mutex lock.
   ///
+  /// ```
   ///   mutex.lock([&value] { value++; });
+  /// ```
   ///
-  /// Precondition: Mutex unlocked by this thread, undefined otherwise.
+  /// Precondition: Mutex not held by this thread, undefined otherwise.
   template <typename CriticalSection>
   void lock(CriticalSection criticalSection) {
```

Comment:
This naming convention should be changed to distinguish simple locking from lock-perform-unlock wrappers. Swift uses `with` for similar functions, which here would be `withLock(...)`, `withReadLock(...)`, etc.